### PR TITLE
[151] Runner executor

### PR DESCRIPTION
**Pull request by giventocode**
Closes: #151 

In this PR:
  -  TES runner performs the execution of commands (executor).
  -  Moves the batch script preparation (creation and uploading) logic outside the BatchScheduler.
     - `TaskExecutionScriptingManager` Manages the creation and uploading of the Batch script and all its dependencies, this is the entry-point class the BatchScheduler uses.
     - `BatchNodeScriptBuilder`. Fluid builder of the script to be executed by Batch. Used by `TaskExecutionScriptingManager`. 
     - `TaskToNodeTaskConverter`  Handles the creation of a NodeTask from a TesTask. Used by `TaskExecutionScriptingManager`. Creation of the NodeTask's inputs and outputs has the following key functionality:
        -  Handles content inputs by creating them as blobs in storage and setting the NodeTasks' inputs to the blobs' URLs.
        - Handles local file paths in inputs when Cromwell is using the local filesystem with a blob FUSE driver.   

     - `NodeTaskBuilder` Fluid builder of NodeTask instances. Contains the logic that sets the URL transformation strategy for inputs/outputs using the configuration settings. Used by `TaskToNodeTaskConverter`.
  - Added new properties to the NodeTask model:
    - `id` and `workflowId`, uniquely identifies the task and its associated workflow id. 
    - Renamed `SasTransformationStrategy` property in `FileOutput` to `TransformationStrategy`
  - Logging entries and formatting changes to facilitate troubleshooting.
  - Added support for local path mapping to URL for Cromwell local directories (running with blob FUSE driver) and configured external storage. accounts
  - I added the `RuntimeOptions.NodeManagedIdentityResourceId` (workflow ID or global ID as fallback) to the node task and changed the runner to use it `ManagedIdentityCredential(new ResourceIdentifier(runtimeOptions.NodeManagedIdentityResourceId));`
If NodeManagedIdentiyResourceId is not set, the runner will use `DefaultAzureCredentials()` as a fallback, which works when identity is assigned and available.
The TES server now uploads the TES task to storage account.
 - Implemented retries in the runner:
    - When downloading docker images
    - When CredentialUnavailableException is thrown while getting the auth token to generate SAS tokens.
 - Known issues:
   -  DRS URIs are no longer supported, depends on #150 
   -  Private container registry functionality is not supported, to be implemented in #279 
   - Docker image clean up is not supported, to be implemented in #363

### CODE REVIEW

File: src/TesApi.Web/BatchNodeScriptBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        private const string BatchTaskDirEnvVar = $"${BatchTaskDirEnvVarName}";
+
+        /// <summary>
+        /// Name of the TES runner binary file
+        /// </summary>
+        public const string NodeTaskRunnerFilename = "tRunner";
+        /// <summary>
+        /// Name of the TES runner task definition file
+        /// </summary>
+        public const string NodeRunnerTaskDefinitionFilename = "TesTask.json";
```
Comment (MattMcL4475) [inline src/TesApi.Web/BatchNodeScriptBuilder.cs:30]: Can we change this name?  Confusing because it's not a Tes.Models.TesTask .  Maybe "NodeTask.json"

Comment (giventocode) [reply src/TesApi.Web/BatchNodeScriptBuilder.cs:30]: Yes, I am planning to do file renaming as a follow item, will include changing the binary filename as well as this file.

File: src/TesApi.Web/BatchNodeScriptBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        /// <param name="setExecutable">Whether the file should be made executable or not</param>
+        /// <returns>The command to execute</returns>
+        public static string CreateWgetDownloadCommand(string urlToDownload, string localFilePathDownloadLocation, bool setExecutable = false)
+        {
+            if (!Uri.TryCreate(urlToDownload, UriKind.Absolute, out var _))
+            {
+                throw new ArgumentException($"Invalid URL: {urlToDownload}", nameof(urlToDownload));
+            }
+
+            string command = $"wget --https-only --timeout=20 --waitretry=1 --tries=9 --retry-connrefused --continue -O {localFilePathDownloadLocation} '{urlToDownload}'";
```
Comment (BMurri) [inline src/TesApi.Web/BatchNodeScriptBuilder.cs:158]: Add `--no-verbose` to wget command-line

File: src/TesApi.Web/BatchScheduler.cs [hunk 3]
```diff
@@ -977,6 +985,142 @@ private ValueTask<bool> HandleTesTaskTransitionAsync(TesTask tesTask, CombinedBa
...
+        {
+            ValidateTesTask(task);
+
+            var additionalInputs = await GetAdditionalInputsAsync(task, cancellationToken);
+
+            // TODO: Confirm if the mapping here is correct. This is a carry over from the previous implementation. 
+            var poolHasContainerConfig = !(isPublic.ExecutorImage && isPublic.DockerInDockerImage && isPublic.CromwellDrsImage);
+            var containerCleanupOptions = new RuntimeContainerCleanupOptions(
+                ExecuteDockerRmi: !poolHasContainerConfig,
+                ExecuteDockerPrune: enableBatchAutopool);
```
Comment (BMurri) [inline src/TesApi.Web/BatchScheduler.cs:1000]: should be `!enableBatchAutopool`. But as discussed, replace both this and the above line with `true`

File: src/TesApi.Web/Runner/NodeTaskBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        /// </summary>
+        /// <param name="nodeTask"></param>
+        public NodeTaskBuilder(NodeTask nodeTask)
+        {
+            this.nodeTask = nodeTask;
+        }
+
+        private void SetDefaultValues()
+        {
+            nodeTask.ContainerWorkDir = "/";
```
Comment (BMurri) [inline src/TesApi.Web/Runner/NodeTaskBuilder.cs:40]: If the container's workdir is not explicitly provided in the TesTask, we don't set it in the docker command-line (or API) (we used to set it to "/", but that was changed over a month ago).

File: src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+            try
+            {
+                var builder = new NodeTaskBuilder();
+
+                //TODO: Revise this assumption (carried over from the current implementation) and consider Single() if in practice only one executor per task is supported.
+                var executor = task.Executors.First();
+
+                builder.WithId(task.Id)
+                    .WithWorkflowId(task.WorkflowId)
+                    .WithContainerCommands(executor.Command.Select(EscapeBashArgument).ToList())
```
Comment (BMurri) [inline src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs:87]: `EscapeBashArgument` is required when each part of the `Command` array from the `TesTask` ends up interpreted by `bash`. If we pass the array directly to the Docker API (which we do here), then they should NOT be escaped.

File: src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            return uri.GetLeftPart(UriPartial.Path);
+        }
+
+        private static string EscapeBashArgument(string arg)
+        {
+            return $"{arg.Replace(@"'", @"'\''")}";
+        }
+
+
+        private static void MapOutputs(List<TesOutput> outputs, string pathParentDirectory, string containerMountParentDirectory,
```
Comment (BMurri) [inline src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs:258]: This needs to convert file schema urls to storage account urls

File: src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+        {
+            outputs?.ForEach(output =>
+            {
+                builder.WithOutputUsingCombinedTransformationStrategy(
+                    AppendParentDirectoryIfSet(output.Path, pathParentDirectory), output.Url, ToNodeTaskFileType(output.Type),
+                    containerMountParentDirectory);
+            });
+        }
+
+        private static void MapInputs(List<TesInput> inputs, string pathParentDirectory, string containerMountParentDirectory,
```
Comment (BMurri) [inline src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs:269]: Same as MapOutputs

**Review by BMurri — CHANGES_REQUESTED** [review]

File: src/TesApi.Web/Storage/DefaultStorageAccessProvider.cs [hunk 8]
```diff
@@ -166,6 +167,24 @@ private static string NormalizedBlobPath(string blobPath)
...
+            var blobPathWithPrefix = $"{task.Id}{normalizedBlobPath}";
+            if (task.Resources?.ContainsBackendParameterValue(TesResources.SupportedBackendParameters
+                    .internal_path_prefix) == true)
+            {
+                blobPathWithPrefix =
+                    $"{task.Resources.GetBackendParameterValue(TesResources.SupportedBackendParameters.internal_path_prefix).Trim('/')}{normalizedBlobPath}";
+            }
+
+            //passing the resulting string through the builder to ensure that the path is properly encoded and valid
+            var builder = new BlobUriBuilder(new Uri($"https://{defaultStorageAccountName}.blob.core.windows.net/{blobPathWithPrefix.TrimStart('/')}"));
```
Comment (MattMcL4475) [inline src/TesApi.Web/Storage/DefaultStorageAccessProvider.cs:183]: is this missing the container name?

File: src/TesApi.Web/Startup.cs [hunk 9]
```diff
@@ -107,6 +108,9 @@ public void ConfigureServices(IServiceCollection services)
                     .AddSingleton<ConfigurationUtils>()
                     .AddSingleton<IAllowedVmSizesService, AllowedVmSizesService>()
                     .AddSingleton<TokenCredential>(s => new DefaultAzureCredential())
+                    .AddSingleton<TaskToNodeTaskConverter>()
+                    .AddSingleton<TaskExecutionScriptingManager>()
+                    .AddTransient<BatchNodeScriptBuilder>()
```
Comment (MattMcL4475) [inline src/TesApi.Web/Startup.cs:113]: Isn't this a utility class?  Is it really appropriate for dependency injection?

File: src/TesApi.Web/Storage/DefaultStorageAccessProvider.cs (hunk 8 above)
Comment (giventocode) [reply src/TesApi.Web/Storage/DefaultStorageAccessProvider.cs:183]: Yes, good catch. This method is not used. I will clean this up.

File: src/TesApi.Web/Startup.cs (hunk 9 above)
Comment (giventocode) [reply src/TesApi.Web/Startup.cs:113]: hmm, not sure I agree. These are dependencies that are required to be injected and their creation requires options and other dependencies from the app context.  I think DI is doing what is supposed to be do.

File: src/Tes.Runner/Docker/DockerExecutor.cs [hunk 10]
```diff
@@ -61,16 +61,17 @@ private async Task<MultiplexedStream> StreamStdOutAndErrorAsync(string container
...
-        private async Task<CreateContainerResponse> CreateContainerAsync(string imageName, List<string> commandsToExecute, List<string>? volumeBindings)
+        private async Task<CreateContainerResponse> CreateContainerAsync(string imageName, string imageTag,
+            List<string> commandsToExecute, List<string>? volumeBindings, string? workingDir)
         {
             var createResponse = await dockerClient.Containers.CreateContainerAsync(
                 new CreateContainerParameters
                 {
-                    Image = imageName,
+                    Image = ToImageNameWithTag(imageName, imageTag),
                     Entrypoint = commandsToExecute,
```
[the remark this answers is not in the document]
Comment (MattMcL4475) [reply src/Tes.Runner/Docker/DockerExecutor.cs]: we set it to an empty string

**Review by MattMcL4475 — APPROVED** [review]

File: src/CommonUtilities/Models/NodeTask.cs [hunk 11]
```diff
@@ -42,7 +42,7 @@ public class RuntimeOptions
     {
         public TerraRuntimeOptions? Terra { get; set; }
 
-        public DockerCleanUpOptions DockerCleanUp { get; set; }
+        public string? NodeManagedIdentityResourceId { get; set; }
     }
 
     public class DockerCleanUpOptions
```
Comment (BMurri) [inline src/CommonUtilities/Models/NodeTask.cs:48]: If this is no longer used, please remove.

File: src/Tes.Runner/Authentication/CredentialsManager.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            var duration = TimeSpan.FromSeconds(Math.Pow(ExponentialBackOffExponent, attempt));
+            logger.LogInformation($"Waiting {duration} before retrying");
+            return duration;
+        }
+
+        public TokenCredential GetTokenCredential(RuntimeOptions runtimeOptions)
+        {
+            return retryPolicy.Execute(() => GetTokenCredentialImpl(runtimeOptions));
+        }
+        private TokenCredential GetTokenCredentialImpl(RuntimeOptions runtimeOptions)
```
Comment (BMurri) [inline src/Tes.Runner/Authentication/CredentialsManager.cs:42]: Need a blank line here

File: src/TesApi.Web/BatchScheduler.cs [hunk 13]
```diff
@@ -169,8 +154,7 @@ public partial class BatchScheduler : IBatchScheduler
                 _batchPoolFactory = poolFactory;
                 batchPrefix = batchSchedulingOptions.Value.Prefix;
                 logger.LogInformation("BatchPrefix: {BatchPrefix}", batchPrefix);
-                taskRunScriptContent = File.ReadAllLines(Path.Combine(AppContext.BaseDirectory, "scripts/task-run.sh"));
-                taskCleanupScriptContent = File.ReadAllLines(Path.Combine(AppContext.BaseDirectory, "scripts/clean-executor.sh"));
+                File.ReadAllLines(Path.Combine(AppContext.BaseDirectory, "scripts/task-run.sh"));
```
Comment (BMurri) [inline src/TesApi.Web/BatchScheduler.cs:157]: If this isn't storing the contents of the file anywhere, this line can go away. If that is the case, please also delete the two files in the `scripts` directory.

File: src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+            {
+                return default;
+            }
+
+            logger.LogInformation($"The input is content. Uploading its content to the internal storage location. Input path:{input.Path}");
+
+            return await UploadContentAndCreateTesInputAsync(tesTask, input.Path, input.Content, cancellationToken);
+        }
+
+        private static string EscapeBashArgument(string arg)
```
Comment (BMurri) [inline src/TesApi.Web/Runner/TaskToNodeTaskConverter.cs:352]: Are you dumping the commands from the TesTask into a script in a single string or are you passing them to the Docker API? If the latter, we should NOT be shell escaping them. The Docker API does not turn them into a single string command-line, it passes them to the process's main() function directly.

**Review by BMurri — APPROVED** [review]
Mostly nitpicks and cleanup. Please open issues for any you don't do now.

File: src/CommonUtilities/Models/NodeTask.cs (hunk 11 above)
Comment (BMurri) [reply src/CommonUtilities/Models/NodeTask.cs:48]: Should have been line 48

**Pull request merged**